File: 0114-flatten-binary-tree-to-linked-list/0114-flatten-binary-tree-to-linked-list.java
class Solution {
    TreeNode prev = null;

    public void flatten(TreeNode root) {
        if (root == null) return;

        // Reverse preorder: right -> left -> root
        flatten(root.right);
        flatten(root.left);

        // Link current node to previously visited node
        root.right = prev;
        root.left = null;
        prev = root;
    }
}

File: 0114-flatten-binary-tree-to-linked-list/0114-flatten-binary-tree-to-linked-list.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

struct TreeNode* prev = NULL;

void solve(struct TreeNode* root) {
    if (root == NULL) return;

    // Save original children before modifying anything
    struct TreeNode* leftChild = root->left;
    struct TreeNode* rightChild = root->right;

    // Link previous node to current
    if (prev != NULL) {
        prev->right = root;
        prev->left = NULL;
    }

    // Move prev to current
    prev = root;

    // Continue preorder traversal
    solve(leftChild);
    solve(rightChild);
}

void flatten(struct TreeNode* root) {
    prev = NULL;      // reset global pointer
    solve(root);
}
